libyul/backends/wasm/EWasmCodeTransform.cpp

```diff
@@ -0,0 +1,317 @@
+{
+	yulAssert(false, "");
+	return {};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(FunctionalInstruction const&)
+{
+	yulAssert(false, "");
+	return {};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(FunctionCall const& _call)
+{
+	if (m_dialect.builtin(_call.functionName.name))
+		return wasm::BuiltinCall{_call.functionName.name.str(), visit(_call.arguments)};
+	else
+		// If this function returns multiple values, then the first one will
+		// be returned in the expression itself and the others in global variables.
+		// The values have to be used right away in an assignment or variable declaration,
+		// so it is handled there.
+		return wasm::FunctionCall{_call.functionName.name.str(), visit(_call.arguments)};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(Identifier const& _identifier)
+{
+	return wasm::LocalVariable{_identifier.name.str()};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(Literal const& _literal)
+{
+	u256 value = valueOfLiteral(_literal);
+	yulAssert(value <= numeric_limits<uint64_t>::max(), "");
+	return wasm::Literal{uint64_t(value)};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(yul::Instruction const&)
+{
+	yulAssert(false, "EVM instruction used for Wasm code generation.");
+	return {};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(If const& _if)
+{
+	return wasm::If{visit(*_if.condition), visit(_if.body.statements), {}};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(Switch const& _switch)
+{
+	wasm::Block block;
+	string condition = m_nameDispenser.newName("condition"_yulstring).str();
+	m_localVariables.emplace_back(wasm::VariableDeclaration{condition});
+	block.statements.emplace_back(wasm::LocalAssignment{condition, visit(*_switch.expression)});
+
+	vector<wasm::Expression>* currentBlock = &block.statements;
+	for (size_t i = 0; i < _switch.cases.size(); ++i)
+	{
+		Case const& c = _switch.cases.at(i);
+		if (c.value)
+		{
+			wasm::BuiltinCall comparison{"i64.eq", {}};
+			comparison.arguments.emplace_back(wasm::LocalVariable{condition});
+			comparison.arguments.emplace_back(visitReturnByValue(*c.value));
+			wasm::If ifStmnt{
+				make_unique<wasm::Expression>(std::move(comparison)),
+				visit(c.body.statements),
+				{}
+			};
+			vector<wasm::Expression>* nextBlock = nullptr;
+			if (i != _switch.cases.size() - 1)
+			{
+				ifStmnt.elseStatements = make_unique<vector<wasm::Expression>>();
+				nextBlock = ifStmnt.elseStatements.get();
+			}
+			currentBlock->emplace_back(std::move(ifStmnt));
+			currentBlock = nextBlock;
+		}
+		else
+		{
+			yulAssert(i == _switch.cases.size() - 1, "Default case must be last.");
+			*currentBlock += visit(c.body.statements);
+		}
+	}
+	return std::move(block);
+}
+
+wasm::Expression EWasmCodeTransform::operator()(FunctionDefinition const&)
+{
+	yulAssert(false, "Should not have visited here.");
+	return {};
+}
+
+wasm::Expression EWasmCodeTransform::operator()(ForLoop const& _for)
+{
```

Comment:
Probably want to generate code for `_for.pre` or assert that `_for.pre` is empty?

Reply:
Ah thanks a lot!